### DISCUSSION TITLE
Cycle over rate exceeds 0.3. Stopping realtime loop...

### DISCUSSION THREAD

**Discussion by sthoduka** [HumanSupportRobot/QandA#86] [Q&A] [answered]
In the last few weeks, we have the following problem:
1. HSR is started
2. After a few minutes (sometimes 2 minutes, sometimes 10 minutes, sometimes more), it says 'HSR stopped', and needs to be restarted (with e-stop press/release).

The problem occurs even if we are not running any functionalities (i.e. just start the robot, and don't launch any other nodes).

The following messages are seen in /rosout when this happens:

```
[ WARN] [1698675773.351875189]: Cycle over rate exceeds 0.1.
[ WARN] [1698675773.846022088]: Cycle over rate exceeds 0.1.
[ERROR] [1698675774.339749818]: Cycle over rate exceeds 0.3. Stopping realtime loop...
```
and the `hsrb/robot_hardware` node dies, which is what causes 'HSR stopped'.

As suggested [here](https://github.com/orgs/HumanSupportRobot/discussions/21), we have tried to disable marker recognition, and several of the sensors, but this does not seem to fix the problem. CPU load has not changed significantly in the last few weeks. 

Has anyone else had this problem? Any other ideas for how to debug this?

**Comment by yuka-hashiguchi** [comment]
Do you use the HSR with the wireless stop button? The error is generated when communication with the amp is interrupted for a certain period due to some factor, and it also occurs when you press the stop button. Since the wireless stop button uses radio waves in the 2.4GHz frequency band, interference with radio waves or a decrease in the button's battery level can cause the problem you mentioned. Does the battery level of the wireless stop button seem to be fine?

**Reply by sthoduka** [reply]
We have actually disabled the wireless e-stop with the  'Wireless emergency stop changeover switch' a couple of years ago because of exactly the problem you described. During our debugging in the last few weeks, we had also re-enabled it, and kept the wireless e-stop fully charged near the robot, but it still eventually died. 

Do you mean that the "Cycle over rate exceeds 0.3. Stopping realtime loop..." error happens only when the e-stop is triggered somehow? Anyway, we will debug a bit more in that direction to see if we find out something. Thanks a lot!

**Reply by yuka-hashiguchi** [reply]
Thank you for providing the information on the situation. When this issue occurred, what color is the HSR status indicator LED illuminated or flashing? If it's flashing, it indicates that the amplifier is not powered on, which could be a hardware-related problem.

**Reply by sthoduka** [reply]
When the error occurs, the status LED illuminates red (not flashing). The motors are powered since we see that the head does not fall backwards, and the wrist remains upright.

**Reply by yuka-hashiguchi** [reply]
Sorry for my late reply. It seems that the amplifier's power is on because the status LED is not flashing. It appears that diagnostics reports some errors. Could you please check from rqt where the error is being output and let me know?

https://docs.hsr.io/hsrc_user_manual_en/troubleshoot/troubleshoot.html?highlight=rqt#diagnostics-viewer

**Reply by sthoduka** [reply]
thanks :)
Here is a screenshot of the diagnostics viewer immediately after the error occurs. Before the error occurs, there are no error messages, and all devices are "OK". 

![rqt_diagnostics](https://github.com/HumanSupportRobot/QandA/assets/5426162/8a43ba8f-806e-493a-baba-488af81b274f)


We also tried the following:
1. Disabled automatic start up (https://docs.hsr.io/hsr_develop_manual_en/customize/automatic_start.html#off)
2. Launch the nodes manually with: `roslaunch hsrb_robot_launch hsrb_boot_all.launch`

The terminal output is as follows:
```
699279776.446631193, when looking up transform from frame [base_footprint] to frame [odom]
[ WARN] [1699279776.805769760]: Could not get base to fixed frame transform, Lookup would require extrapolation 0.099950253s into the future.  Requested time 1699279776.672328711 but the latest data is at time 1699279776.572378397, when looking up transform from frame [base_footprint] to frame [odom]
[ WARN] [1699279776.907427092]: Cycle over rate exceeds 0.1.
[ WARN] [1699279776.909902062]: Could not get base to fixed frame transform, Lookup would require extrapolation 0.124604362s into the future.  Requested time 1699279776.796933174 but the latest data is at time 1699279776.672328711, when looking up transform from frame [base_footprint] to frame [odom]
[ WARN] [1699279777.014331280]: Could not get base to fixed frame transform, Lookup would require extrapolation 0.100922607s into the future.  Requested time 1699279776.897855759 but the latest data is at time 1699279776.796933174, when looking up transform from frame [base_footprint] to frame [odom]
[ WARN] [1699279777.119778377]: Could not get base to fixed frame transform, Lookup would require extrapolation 0.074569897s into the future.  Requested time 1699279776.972425699 but the latest data is at time 1699279776.897855759, when looking up transform from frame [base_footprint] to frame [odom]
[ WARN] [1699279777.223907814]: Could not get base to fixed frame transform, Lookup would require extrapolation 0.100140195s into the future.  Requested time 1699279777.072565794 but the latest data is at time 1699279776.972425699, when looking up transform from frame [base_footprint] to frame [odom]
[ WARN] [1699279777.328090472]: Could not get base to fixed frame transform, Lookup would require extrapolation 0.150014799s into the future.  Requested time 1699279777.222580671 but the latest data is at time 1699279777.072565794, when looking up transform from frame [base_footprint] to frame [odom]
[ WARN] [1699279777.405986441]: Cycle over rate exceeds 0.1.
[ WARN] [1699279777.434184857]: Could not get base to fixed frame transform, Lookup would require extrapolation 0.074776825s into the future.  Requested time 1699279777.297357559 but the latest data is at time 1699279777.222580671, when looking up transform from frame [base_footprint] to frame [odom]
[ WARN] [1699279777.538991379]: Could not get base to fixed frame transform, Lookup would require extrapolation 0.100524053s into the future.  Requested time 1699279777.397881508 but the latest data is at time 1699279777.297357559, when looking up transform from frame [base_footprint] to frame [odom]
[ WARN] [1699279777.642869233]: Could not get base to fixed frame transform, Lookup would require extrapolation 0.124849290s into the future.  Requested time 1699279777.522730827 but the latest data is at time 1699279777.397881508, when looking up transform from frame [base_footprint] to frame [odom]
[ WARN] [1699279777.747185587]: Could not get base to fixed frame transform, Lookup would require extrapolation 0.100083003s into the future.  Requested time 1699279777.622813702 but the latest data is at time 1699279777.522730827, when looking up transform from frame [base_footprint] to frame [odom]
[ WARN] [1699279777.853821287]: Could not get base to fixed frame transform, Lookup would require extrapolation 0.074887712s into the future.  Requested time 1699279777.697701454 but the latest data is at time 1699279777.622813702, when looking up transform from frame [base_footprint] to frame [odom]
[ERROR] [1699279777.904317082]: Cycle over rate exceeds 0.3. Stopping realtime loop...
[ WARN] [1699279777.958029200]: Could not get base to fixed frame transform, Lookup would require extrapolation 0.150460161s into the future.  Requested time 1699279777.848161697 but the latest data is at time 1699279777.697701454, when looking up transform from frame [base_footprint] to frame [odom]
[ WARN] [1699279778.062460254]: Could not get base to fixed frame transform, Lookup would require extrapolation 0.074949238s into the future.  Requested time 1699279777.923110962 but the latest data is at time 1699279777.848161697, when looking up transform from frame [base_footprint] to frame [odom]
[ WARN] [1699279778.166239437]: Could not get base to fixed frame transform, Lookup would require extrapolation 0.100037815s into the future.  Requested time 1699279778.023148775 but the latest data is at time 1699279777.923110962, when looking up transform from frame [base_footprint] to frame [odom]
[ WARN] [1699279778.272491885]: Could not get base to fixed frame transform, Lookup would require extrapolation 0.125493749s into the future.  Requested time 1699279778.148642540 but the latest data is at time 1699279778.023148775, when looking up transform from frame [base_footprint] to frame [odom]
[ WARN] [1699279778.376574741]: Could not get base to fixed frame transform, Lookup would require extrapolation 0.099713865s into the future.  Requested time 1699279778.248356342 but the latest data is at time 1699279778.148642540, when looking up transform from frame [base_footprint] to frame [odom]
[ WARN] [1699279778.480230514]: Could not get base to fixed frame transform, Lookup would require extrapolation 0.100275615s into the future.  Requested time 1699279778.348631859 but the latest data is at time 1699279778.248356342, when looking up transform from frame [base_footprint] to frame [odom]
[ WARN] [1699279778.584205706]: Could not get base to fixed frame transform, Lookup would require extrapolation 0.124947977s into the future.  Requested time 1699279778.473579884 but the latest data is at time 1699279778.348631859, when looking up transform from frame [base_footprint] to frame [odom]
[hsrb/robot_hardware-1] process has finished cleanly
log file: /home/sthoduka/.ros/log/20231106-133751_aebefafe-7ca9-11ee-8802-0030644a5121/hsrb-robot_hardware-1*.log

```

Filtering out the TF warnings, the error messages are the same as those I mentioned in the initial post:
```
[ WARN] [1699279776.907427092]: Cycle over rate exceeds 0.1.
[ WARN] [1699279777.405986441]: Cycle over rate exceeds 0.1.
[ERROR] [1699279777.904317082]: Cycle over rate exceeds 0.3. Stopping realtime loop...
[hsrb/robot_hardware-1] process has finished cleanly
log file: /home/sthoduka/.ros/log/20231106-133751_aebefafe-7ca9-11ee-8802-0030644a5121/hsrb-robot_hardware-1*.log
```

From the dashboard, there is some additional info regarding the location of the error

```
Node: /hsrb/robot_hardware

Time: 1696606497.330406264

Severity: Error

Location: /build/ros-noetic-tmc-realtime-tools-0.9.3/src/tmc_realtime_tools/run_loop.cpp:in `{anonymous}::RunLoopImpl':204

Published Topics: /rosout, /hsrb/joint_states, /hsrb/servo_states, /hsrb/omni_base_controller/state, /hsrb/omni_base_controller/internal_state, /hsrb/wheel_odom, /tf, /hsrb/omni_base_controller/follow_joint_trajectory/result, /hsrb/omni_base_controller/follow_joint_trajectory/feedback, /hsrb/omni_base_controller/follow_joint_trajectory/status, /hsrb/head_trajectory_controller/state, /hsrb/head_trajectory_controller/follow_joint_trajectory/result, /hsrb/head_trajectory_controller/follow_joint_trajectory/feedback, /hsrb/head_trajectory_controller/follow_joint_trajectory/status, /hsrb/arm_trajectory_controller/state, /hsrb/arm_trajectory_controller/follow_joint_trajectory/result, /hsrb/arm_trajectory_controller/follow_joint_trajectory/feedback, /hsrb/arm_trajectory_controller/follow_joint_trajectory/status, /hsrb/drive_mode, /hsrb/gripper_controller/follow_joint_trajectory/result, /hsrb/gripper_controller/follow_joint_trajectory/feedback, /hsrb/gripper_controller/follow_joint_trajectory/status, /hsrb/gripper_controller/grasp/result, /hsrb/gripper_controller/grasp/feedback, /hsrb/gripper_controller/grasp/status, /hsrb/gripper_controller/apply_force/result, /hsrb/gripper_controller/apply_force/feedback, /hsrb/gripper_controller/apply_force/status, /hsrb/base_imu/data, /hsrb/overload_joint_state, /hsrb/collision_detection_controller/drivers/arm_flex_joint/settings/Current/parameter_descriptions, /hsrb/collision_detection_controller/drivers/arm_flex_joint/settings/Current/parameter_updates, /hsrb/collision_detection_controller/drivers/arm_flex_joint/settings/Effort/parameter_descriptions, /hsrb/collision_detection_controller/drivers/arm_flex_joint/settings/Effort/parameter_updates, /hsrb/collision_detection_controller/drivers/arm_lift_joint/settings/Current/parameter_descriptions, /hsrb/collision_detection_controller/drivers/arm_lift_joint/settings/Current/parameter_updates, /hsrb/collision_detection_controller/drivers/arm_lift_joint/settings/Effort/parameter_descriptions, /hsrb/collision_detection_controller/drivers/arm_lift_joint/settings/Effort/parameter_updates, /hsrb/collision_detection_controller/drivers/arm_roll_joint/settings/Current/parameter_descriptions, /hsrb/collision_detection_controller/drivers/arm_roll_joint/settings/Current/parameter_updates, /hsrb/collision_detection_controller/drivers/arm_roll_joint/settings/Effort/parameter_descriptions, /hsrb/collision_detection_controller/drivers/arm_roll_joint/settings/Effort/parameter_updates, /hsrb/collision_detection_controller/drivers/base_l_drive_wheel_joint/settings/Current/parameter_descriptions, /hsrb/collision_detection_controller/drivers/base_l_drive_wheel_joint/settings/Current/parameter_updates, /hsrb/collision_detection_controller/drivers/base_l_drive_wheel_joint/settings/Effort/parameter_descriptions, /hsrb/collision_detection_controller/drivers/base_l_drive_wheel_joint/settings/Effort/parameter_updates, /hsrb/collision_detection_controller/drivers/base_roll_joint/settings/Current/parameter_descriptions, /hsrb/collision_detection_controller/drivers/base_roll_joint/settings/Current/parameter_updates, /hsrb/collision_detection_controller/drivers/base_roll_joint/settings/Effort/parameter_descriptions, /hsrb/collision_detection_controller/drivers/base_roll_joint/settings/Effort/parameter_updates, /hsrb/collision_detection_controller/drivers/base_r_drive_wheel_joint/settings/Current/parameter_descriptions, /hsrb/collision_detection_controller/drivers/base_r_drive_wheel_joint/settings/Current/parameter_updates, /hsrb/collision_detection_controller/drivers/base_r_drive_wheel_joint/settings/Effort/parameter_descriptions, /hsrb/collision_detection_controller/drivers/base_r_drive_wheel_joint/settings/Effort/parameter_updates, /hsrb/collision_detection_controller/drivers/head_pan_joint/settings/Current/parameter_descriptions, /hsrb/collision_detection_controller/drivers/head_pan_joint/settings/Current/parameter_updates, /hsrb/collision_detection_controller/drivers/head_pan_joint/settings/Effort/parameter_descriptions, /hsrb/collision_detection_controller/drivers/head_pan_joint/settings/Effort/parameter_updates, /hsrb/collision_detection_controller/drivers/head_tilt_joint/settings/Current/parameter_descriptions, /hsrb/collision_detection_controller/drivers/head_tilt_joint/settings/Current/parameter_updates, /hsrb/collision_detection_controller/drivers/head_tilt_joint/settings/Effort/parameter_descriptions, /hsrb/collision_detection_controller/drivers/head_tilt_joint/settings/Effort/parameter_updates, /hsrb/collision_detection_controller/drivers/wrist_flex_joint/settings/Current/parameter_descriptions, /hsrb/collision_detection_controller/drivers/wrist_flex_joint/settings/Current/parameter_updates, /hsrb/collision_detection_controller/drivers/wrist_flex_joint/settings/Effort/parameter_descriptions, /hsrb/collision_detection_controller/drivers/wrist_flex_joint/settings/Effort/parameter_updates, /hsrb/collision_detection_controller/drivers/wrist_roll_joint/settings/Current/parameter_descriptions, /hsrb/collision_detection_controller/drivers/wrist_roll_joint/settings/Current/parameter_updates, /hsrb/collision_detection_controller/drivers/wrist_roll_joint/settings/Effort/parameter_descriptions, /hsrb/collision_detection_controller/drivers/wrist_roll_joint/settings/Effort/parameter_updates, /hsrb/diagnostics

Cycle over rate exceeds 0.3. Stopping realtime loop...
```

**Comment by yoshimi-iyoda** [comment]
I would like to check the logs and would appreciate it if you could collect the logs with hsrb_command using the following as a reference.
https://docs.hsr.io/hsrb_user_manual_en/troubleshoot/troubleshoot.html#upload-log-tool



Requests for log collection

- Please log collection of when the error occurred and also let me know the exact UTC time at that time.
- The range of the log collection should be about 1 day before or after the date of the error.
- Log collection when the wireless function was disabled.
- Please paste the Sentence line to the result of log collection tool ([LOG] robot_log_....)

**Reply by sthoduka** [reply]
> Log collection when the wireless function was disabled.

Do you mean the wireless e-stop should be disabled?

**Reply by sthoduka** [reply]
Thank you for the instructions.

I have completed the log collection, starting from 2023/11/06 until 2023/11/07 (around 9:25 UTC). 

The UTC time when the error occurred is Tue Nov  7 09:21:07 UTC 2023 (or epoch timestamp 1699348867).

The sentence line is as follows:

[LOG] robot_log_000c8b200889/20231107_102326_log.zip

**Reply by yoshimi-iyoda** [reply]
Thank you for collecting the logs.

> Do you mean the wireless e-stop should be disabled?

Sorry for the lack of explanation,
I dared to write this because I also want to compare the logs when it is working properly (when the wireless e-stop function is disabled) and when it causes an error (when the wireless e-stop function is enabled).

First, we will check the logs you collected.

**Reply by yoshimi-iyoda** [reply]
I checked and it was only logging up to 2023/11/06 14:28 (UTC) as shown below, perhaps due to a problem with hsrb_command.

```
journalctl --file=hsrb70*
-- Logs begin at Mon 2023-11-06 23:00:57 JST, end at Mon 2023-11-06 23:28:50 JST.

```
```
$ tail -2 rosout.log
1699280929.465353295 WARN /move_group.....
```


Could you please run the command with a slightly wider scope as follows?

hsrb_command upload_log 2023/11/5-2023/11/8

**Reply by sthoduka** [reply]
I have re-run the logging with the command you suggested. I hope that captured the error from yesterday.

```
[LOG] robot_log_000c8b200889/20231108_093103_log.zip
```

Thanks for the clarification about the wireless e-stop. The error from yesterday (Tue Nov 7 09:21:07 UTC 2023) occurred with the wireless stop disabled.  I will try to capture another log today with the wireless e-stop enabled.

**Reply by sthoduka** [reply]
This time with the wireless e-stop enabled, the error occurred at Wed Nov  8 08:39:39 UTC 2023

I created another log using the command `hsrb_command upload_log 2023/11/6-2023/11/9`. In case it did not capture the error, I can run the log again tomorrow.

```
[LOG] robot_log_000c8b200889/20231108_100536_log.zip
```

**Reply by yoshimi-iyoda** [reply]
Thank you for collecting the logs.
 `robot_log_000c8b200889/20231108_093103_log.zip`contained the log of the problem.
Please give me some time to investigate.

**Reply by yoshimi-iyoda** [reply]
We have checked your logs and have not been able to determine the cause of the problem.
We will discuss this with TME representative and get back to you, so please bear with us for a moment.

The following is what we found from the logs.

- No problems were found with CPU usage, memory, IO, etc.
- Error occurred when the HSR was not moving.
- The servo was turned on when the error occurred, but only communication was not possible.

**Reply by sthoduka** [reply]
thank you for the update!

**Reply by yoshimi-iyoda** [reply]
Sorry for the delay.
We have checked the logs and found that there were places where information could not be obtained or the number of power ON/OFF cycles increased for only 21 axes(arm_lift_joint), as shown below.
Possible causes are poor contact of the power supply connector of axis 21, poor contact of the cable, or a problem with the motor amplifier.

First, visually check the power cable and connector of the 21 axes, and unplug and plug the connectors.
The power cable is a black and white twisted connector marked CN6 on the board.
The location is shown in the figure below.



<img width="464" alt="Amplifier21" src="https://github.com/HumanSupportRobot/QandA/assets/71624812/d56bbea2-674a-48dd-aba8-32554b6f7f46">

For instructions on how to remove the cover, please refer to the following
<https://docs.hsr.io/hsrb_maintenance_manual_en/src/remove_cover.html>

Could you please check if this improves the situation?

**Reply by sthoduka** [reply]
Thanks for the instructions. We followed the steps you mentioned, but the same error still occurs.

1. Visual inspection of the cable: there was nothing unusual about the cable or connector. We also followed it (visually) to the other end of the cable to CN242? and did not notice any obvious problems.
2. We unplugged the connector on CN6, and plugged it back in, and put back the cover.
3. Started the robot

The error occurred again after about 25 minutes, and after restarting it, again in around 5 minutes.

One thing we have noticed is that the first time we start up the robot, it takes a longer time (~15-20 minutes) for the error to happen. But after we reset the robot, the error happens in a much shorter time (2-10 minutes). So we were wondering if it could be temperature related.

Could you let us know in which log you found the information regarding axis 21? Maybe we can follow that log the next time the error happens.

**Reply by yoshimi-iyoda** [reply]
Thank you for checking the wiring, connectors, etc.

I found an anomaly in the diag information that is collected when the stop button is released and the amplifier and CPU start communicating.

diag converted /var/log/ros/latest/*bag to csv using a tool.
See below for how to do this.
https://docs.hsr.io/hsr_log_analysis_manual_en/log_files.html#id8


Below are the abnormalities that I found.
Please note that the timestamp is in JST.


Case where communication with axis 21 is not possible and the number of times the power is turned on is unknown.
```
$ cat hsrb_robot_service:_present_log_power_cycle.csv
Timestamp,Level,Message,Hardware ID,11,12,13,21,22,23,24,25,31,32,41
2023/11/08 17:37:02,0,exxx_read present_log_power_cycle succeeded,motor present_log_power_cycle,6926,6926,6925,6942,6925,6925,6924,6923,6926,6929,6930
2023/11/08 17:44:58,0,exxx_read present_log_power_cycle succeeded,motor present_log_power_cycle,6927,6927,6926,6943,6926,6926,6925,6924,6927,6930,6931
2023/11/08 18:03:50,2,exxx_read present_log_power_cycle failed,motor present_log_power_cycle,6929,6929,6928,Unknown,6928,6928,6927,6926,6929,6932,6933
2023/11/08 18:04:54,0,exxx_read present_log_power_cycle succeeded,motor present_log_power_cycle,6929,6929,6928,6945,6928,6928,6927,6926,6929,6932,6933
```


A case where only axis 21 has an increase in the number of power-on times by 4.
```
$ cat hsrb_robot_service:_present_log_power_cycle.csv
Timestamp,Level,Message,Hardware ID,11,12,13,21,22,23,24,25,31,32,41
2023/11/06 22:39:55,0,exxx_read present_log_power_cycle succeeded,motor present_log_power_cycle,6919,6919,6918,6934,6918,6918,6917,6916,6919,6922,6923
2023/11/06 22:59:31,0,exxx_read present_log_power_cycle succeeded,motor present_log_power_cycle,6919,6919,6918,6934,6918,6918,6917,6916,6919,6922,6923
2023/11/06 23:28:11,0,exxx_read present_log_power_cycle succeeded,motor present_log_power_cycle,6921,6921,6920,6938,6920,6920,6919,6918,6921,6924,6925

```

I haven't tried it this time, but you might be able to get some information with this command as well.
For your reference)
https://docs.hsr.io/hsr_faq_en/hsr_common_faq.html#id3

**Reply by yoshimi-iyoda** [reply, marked as answer]
I have asked TME to send a replacement 21-axis amplifier to you.
When the amp arrives, please replace it when you have time.

**Reply by sthoduka** [reply]
We found the logs you mentioned, and also see the extra power cycles for axis 21 in the logs, but did not find anything else new. We will update you once we receive the amplifier and replace it. Thanks!

**Reply by sthoduka** [reply]
We replaced the board with the new one, and performed the [Zero Position Reset Procedure](https://docs.hsr.io/hsrb_maintenance_manual_en/src/reset_0_position_first_aid.html) for axis 21, and it seems to have fixed the problem. We used the HSR for about 30-45 minutes with no issues, so hopefully it means the problem has been resolved. Thanks a lot for the support!

Do you want us to return the old amplifier board?

**Reply by yoshimi-iyoda** [reply]
I have also discussed this with TME and they do not need the old amp, so could you please dispose of it?

**Reply by sthoduka** [reply]
ok sure, we will dispose of it.
